### PR TITLE
Remove GOVUK javascript from buttons

### PR DESCRIPTION
**Pull request by CJWaszczuk**
# What?

Remove GOVUK javascript from buttons

# Why?

GOVUK currently polyfills document within their javascript code for buttons. We may wish to run these components with a server, which means that document would not be available. Therefore, we would either have to polyfill document in other applications or remove the javascript. In this case, we have decided to remove the javascript from GOVUK, so we do not need to polyfill within server side applications using the components, at the cost of having to re-write the javascript implementation.

### CODE REVIEW

**Comment by LBHELewis** [comment]
We also have a cost here that any changes to govuk button javascript will not be picked up here. How necessary is this change?

File: src/components/Button/Button.tsx [hunk 1]
```diff
@@ -54,21 +54,16 @@ export class Button extends React.Component<ButtonProps> {
...
     }
 
-    // `debounce` calls `event.preventDefault()` for us, so we don't need to
-    // repeat that behaviour.
-    const shouldHandleClick = GovukButton.prototype.debounce.call(this, event);
-
-    if (shouldHandleClick === false) {
+    if (debounce(event.target as HTMLButtonElement, this.debounceContext)) {
       return;
     }
```
Comment (erbridge) [inline src/components/Button/Button.tsx]: Shouldn't these `preventDefault`, too?

File: src/components/Button/Button.test.tsx [hunk 2]
```diff
@@ -32,26 +30,34 @@ it("renders correctly without optional props", () => {
...
+  const mockClickEvent = {
+    ...jest.fn()(),
+    target: button.getDOMNode()
+  };
+
+  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  buttonOnClick!.call(button, mockClickEvent);
 
   expect(mockClickHandler).toHaveBeenCalledTimes(1);
-  expect(mockClickHandler).toHaveBeenCalledWith(mockClickEvent);
```
Comment (erbridge) [inline src/components/Button/Button.test.tsx]: Why was this removed?

**Comment by erbridge** [comment]
@LBHELewis `govuk-frontend` doesn't do a good job of detecting if it's in a browser environment or not (and therefore, whether `document` and friends is available), so it throws exceptions when doing server-side rendering. Also, given that `govuk-frontend` doesn't intend for you to call `Button.prototype.debounce` yourself, it's not really outward facing API, so it's implementation could change in a minor or patch version without violating SemVer, potentially breaking our usage in subtle ways in the future. I think it's better to eject from `govuk-frontend`, allowing us to clean up the API a bit, making it easier to use, and to track changes to the behaviour of `govuk-frontend`'s JavaScript, given that the rest of the implementation is only loosely connected to `govuk-frontend`, here. Hopefully their release notes should explain any bugs they fix or improvements they make to user interaction to make that less painful than it might be.

We should probably stop Dependabot from automerging and updates to `lbh-frontend` and `govuk-frontend` to ensure we're confirming version changes. @CJWaszczuk can you add a ticket to the sprint please?

@CJWaszczuk Please add an ADR explaining why we decided to reimplement the JavaScript (and will probably need to for other things, too).

File: src/helpers/debounce.test.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
+import GovukButton from "govuk-frontend/govuk/components/button/button";
+import React from "react";
+
+import { debounce, makeDebounceContext } from "./debounce";
+
+it("returns `false` and doesn't prevent the default action on both of two call within the default timeout without `data-prevent-double-click` set on the event's target, similar to `GovukButton.prototype.debounce`", () => {
```
Comment (CJWaszczuk) [inline src/helpers/debounce.test.ts]: "returns `false` and doesn't prevent the default action on both of **the** two call within the default timeout when `data-prevent-double-click` isn't set on the event's target, similar to `GovukButton.prototype.debounce`"

File: src/helpers/debounce.test.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  {
+    const context = {};
+
+    runTestFor(GovukButton.prototype.debounce.bind(context), undefined);
+  }
+});
+
+it('returns `true` and prevents the default action of the event when called a second time within the default timeout with `data-prevent-double-click` set to `"true"` on the event\'s target, similar to `GovukButton.prototype.debounce`', () => {
```
Comment (CJWaszczuk) [inline src/helpers/debounce.test.ts]: "returns `true` and prevents the default action of the event on a second call within the default timeout when `data-prevent-double-click` set to `"true"` on the event\'s target, similar to `GovukButton.prototype.debounce`'

File: src/helpers/debounce.test.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  {
+    const context = {};
+
+    runTestFor(GovukButton.prototype.debounce.bind(context), undefined, false);
+  }
+});
+
+it("returns `false` and doesn't prevent the default action of the event when called a second time outside of the default timeout with `data-prevent-double-click` set to `\"true\"` on the event's target, similar to `GovukButton.prototype.debounce`", () => {
```
Comment (CJWaszczuk) [inline src/helpers/debounce.test.ts]: 'on a second call outside of the default timeout when' for consistency maybe?

**Review by erbridge — APPROVED** [review]
I'm good to merge this if you are @CJWaszczuk.

**Comment by erbridge** [comment]
@LBHELewis We've added some tests to validate that the behaviour is the same as the `govuk-frontend` version, whish should make it a less risky endeavour.
